### PR TITLE
JBTM 3538 Fix for TCK timeout test failures

### PR DESCRIPTION
**Pull request by mmusgrov**
https://issues.redhat.com/browse/JBTM-3538

This PR is designed to fix timeout test instabilities.
It is in 4 separate, but related, commits. It also reverts the recent change to use the reaper for timeouts since that does not work with WildFly (we can revisit that in a future commit after discussion with the resteasy team).
After fixing the timeout instability I found that 3 of the nested LRA tests failed which I fixed in the last commit.

LRA
!MAIN !CORE !QA_JTA !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !QA_JTS_JACORB !BLACKTIE !XTS !PERF NO_WIN !RTS !AS_TESTS !TOMCAT !JACOCO

### CODE REVIEW

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java [hunk 1]
```diff
@@ -114,11 +116,13 @@ public LRAParticipantRecord() {
                 if (parseException[0] != null) {
                     String errorMsg = LRALogger.i18nLogger.error_invalidCompensator(lra.getId(), parseException[0].getMessage(), linkURI);
                     LRALogger.logger.error(errorMsg);
+                    trace_progress(errorMsg);
```
Comment (ochaloup) [inline rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java:119]: First, I'm fine this being in the code as you prefer it. This is just I would like just express my opinion.
This is kind of like debugging with `System.out.println` and that way it slows down the processing in general. Hopefully the JIT will cover it.

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java [hunk 2]
```diff
@@ -397,8 +410,13 @@ protected ReentrantLock tryLockTransaction() {
     public int finishLRA(boolean cancel) {
         ReentrantLock lock = null;
 
-        // remove the LRA from the timeout scheduler (no need to take the lock)
-        TransactionReaper.transactionReaper().remove(this);
+        // check whether the transaction should cancel due to a timeout:
+        if (finishTime != null && !cancel && ChronoUnit.MILLIS.between(LocalDateTime.now(ZoneOffset.UTC), finishTime) <= 0) {
```
Comment (ochaloup) [inline rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java:414]: This sounds good. As it makes the timeout processing stable it's a great fix.

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java [hunk 3]
```diff
@@ -165,21 +164,6 @@ public Response actionWithLRA(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI contextI
             return getResult(cancel, contextId);
         }
 
-        @GET
-        @Path("time-limit")
```
Comment (ochaloup) [inline rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java:169]: Would you consider to have the test `time-limit` left in the code even there will be no reaper used?

**Review by ochaloup — COMMENTED** [review]
The PR is fine I would say. I have got just few comments.

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java (hunk 1 above)
Comment (mmusgrov) [reply rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java:119]: Your statement is incorrect: trace_progress is coded as `if (LRALogger.logger.isTraceEnabled()) {...}` and I make sure I only call `trace_progress` with static strings from the heap, ie the only performance hit is an extra frame.

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java (hunk 3 above)
Comment (mmusgrov) [reply rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java:169]: That is part of `git revert` for the reaper change.
We have timeout tests in the TCK as well.

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java (hunk 2 above)
Comment (mmusgrov) [reply rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java:414]: Thanks, simple fixes are so much sweater.

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java (hunk 3 above)
Comment (mmusgrov) [reply rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java:169]: But you are right, if I deemed it useful then, then it is still useful. Thanks for picking up on it.

Note that the reason I have included a number of unit tests with the coordinator is becuase they generally complement the TCK and they also directly test coordinator functionality so they are handy to have when making coordinator specific changes.

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java (hunk 1 above)
Comment (ochaloup) [reply rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java:119]: I think my statement is quite correct. The performance hit is an extra frame to the stack and an if operation to verify the trace is enabled. In the hot code path it's not a great strategy. If you stress this as an benefit for other reasons, no problem.

File: rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java (hunk 3 above)
Comment (ochaloup) [reply rts/lra/coordinator/src/test/java/io/narayana/lra/coordinator/domain/model/LRATest.java:169]: +1

**Review by ochaloup — APPROVED** [review]

**Comment by ochaloup** [comment]
The github CI actions fail with `LRATest.mixedMultiLevelNestedActivity:578->multiLevelNestedActivity:706 multiLevelNestedActivity: step 9 (called test path http://localhost:8081/base/test/multiLevelNestedActivity) expected:<1> but was:<2>`
It's a test which happens to start failing quite regurarly on the github CI. I don't know what's the reason for that. Just mentioning, it's some instability

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java (hunk 1 above)
Comment (mmusgrov) [reply rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java:119]: The logical conclusion of your argument is to remove all of our debug and trace level logging and we have plenty of that.

A single frame is minimal and if you really were worried about performance then there are loads of other opportunities to improve the throughput, besides the calls are only ever made with strings from the heap. If we find sufficient uptake of LRA then we can begin looking at sprucing up the code base, until then maintainability is more important.

Is this an impasse for which we need an adjudicator?

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java [hunk 4]
```diff
@@ -1000,4 +1033,16 @@ private String getLRAId(URI lraId) {
...
+        if (LRALogger.logger.isTraceEnabled()) {
+            LRALogger.logger.tracef("%s: Participant id: %s, LRA id: %s, reason: %s, state: %s, accepted: %b",
+                    LocalDateTime.now(ZoneOffset.UTC), // use the same time function as used for LRA timeouts
+                    lraId,
+                    participantPath,
+                    reason,
+                    status,
+                    accepted);
+        }
+    }
```
Comment (jmfinelli) [inline rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java:1047]: WDYT about moving `trace_progress` to `io.narayana.lra.logging`? For example, adding it to `LRALogger` or creating a new utility class?

**Review by jmfinelli — COMMENTED** [review]
LGTM +1

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java (hunk 4 above)
Comment (mmusgrov) [reply rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java:1047]: It's just a trace statement, there is no desire to over-engineer this.

**Review by jmfinelli — APPROVED** [review]

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java (hunk 1 above)
Comment (mmusgrov) [reply rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java:119]: Do also notice that the module is I/O bound and most of the trace logging happens in response to a state change which needs to be persisted, ie the performance hit of an extra frame is dwarfed by other costs.

**Comment by mmusgrov** [comment]
> The github CI actions fail with `LRATest.mixedMultiLevelNestedActivity:578->multiLevelNestedActivity:706 multiLevelNestedActivity: step 9 (called test path http://localhost:8081/base/test/multiLevelNestedActivity) expected:<1> but was:<2>`
> It's a test which happens to start failing quite regurarly on the github CI. I don't know what's the reason for that. Just mentioning, it's some instability

The specification allows the callback to be resent (e.g. to take account of network errors) so the test needs to take account of that (i.e. it should be asserting that the compensate count is >= 1).

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java (hunk 1 above)
Comment (ochaloup) [reply rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java:119]: Sorry for not being clear. This is not an impasse. I'm fine with the PR to be merged and I've approved it already.

Just I have a different perspective on coding practices. For example, I would be supporting and idea to find the places where to emit open tracing events. While I would not track the processing with log trace messages.
The most of the log are written at places of a non standard processing. That's how it should be by me. Not logging on hot code path but in strange and exceptional situations.

Never-mind, I don't want to both you with my arguments. I feel I need to express my perspective which can lead to some discussion on the topic. I don't oppose merging this PR.

**Comment by ochaloup** [comment]
ad `LRATest.mixedMultiLevelNestedActivity:578->multiLevelNestedActivity`) do you plan to put the fix for the test as part of this PR or should I create a new issue for adjusting the test?

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java (hunk 1 above)
Comment (mmusgrov) [reply rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java:119]: Let's wait until we have an opentracing or a bytecode/byteman based solution before throwing out trace logging.
Until that work get's done we need to be able trace the progress of LRAs and compensators.

**Pull request merged**